hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +48,238 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+
+        // check if some of the masses are less or equal to 0
+        bool invalid_mass = false;
+        bool central_interacting = false;
+        for (unsigned int idx = 0; idx < N_tot && !(invalid_mass && central_interacting); ++idx)
+            {
+            // get the index from the embedded group
+            const unsigned int particle_index = h_embed_group.data[idx];
+            // check mass
+            const Scalar4 vel_mass = h_vel_embed.data[particle_index];
+            const Scalar mass = vel_mass.w;
+            if (mass <= Scalar(0))
+                {
+                invalid_mass = true;
+                }
+
+            // check if particle is central particle in rigid body
+            const unsigned int central_tag = h_body_embed.data[particle_index];
+            if (central_tag < MIN_FLOPPY)
+                {
+                const unsigned int central_idx = h_rtag_embed.data[central_tag];
+                if (particle_index == central_idx)
+                    {
+                    central_interacting = true;
+                    }
+                }
+            }
+
+#ifdef ENABLE_MPI
+// MPI_Reduce invalid_mass using logical or if communicating
+#endif // ENABLE_MPI
+
+        if (invalid_mass)
+            {
+            m_exec_conf->msg->warning() << "Some particles have a mass <= 0, may lead to "
+                                           "invalid results during MPCD collision."
+                                        << std::endl;
+            }
+        if (central_interacting)
+            {
+            m_exec_conf->msg->warning() << "Central particle of rigid body included in "
+                                           "MPCD collision. Check if this is intentional."
+                                        << std::endl;
+            }
+        }
+    m_checked_collision_warnings = true;
     }
 
+void mpcd::CollisionMethod::beginRigidBodyCollision(uint64_t timestep)
+    {
+    unsigned int N_tot = m_embed_group->getNumMembers();
+    ArrayHandle<Scalar4> h_initial_vel(m_initial_velocity,
+                                       access_location::host,
+                                       access_mode::overwrite);
+    ArrayHandle<unsigned int> h_embed_group(m_embed_group->getIndexArray(),
+                                            access_location::host,
+                                            access_mode::read);
+    ArrayHandle<Scalar4> h_vel_embed(m_pdata->getVelocities(),
+                                     access_location::host,
+                                     access_mode::read);
+    for (unsigned int idx = 0; idx < N_tot; ++idx)
+        {
+        // collect the initial velocities of the embedded particles
+        const unsigned int particle_index = h_embed_group.data[idx];
+        h_initial_vel.data[idx] = h_vel_embed.data[particle_index];
+        }
+    }
+void mpcd::CollisionMethod::finishRigidBodyCollision(uint64_t timestep)
```

Comment:
Let's have a quick discussion offline on how to organize this code a bit to make it more readable / use scope of `ArrayHandle`s correctly. We can do this during our project meeting if you're OK with that.
```suggestion

void mpcd::CollisionMethod::finishRigidBodyCollision(uint64_t timestep)
```

Reply:
Sure!